flink-connectors/flink-connector-kafka-base/src/main/java/org/apache/flink/streaming/util/serialization/AvroRowSerializationSchema.java

```diff
@@ -0,0 +1,122 @@
+import org.apache.avro.io.Encoder;
+import org.apache.avro.io.EncoderFactory;
+import org.apache.avro.reflect.ReflectDatumWriter;
+import org.apache.avro.specific.SpecificData;
+import org.apache.avro.specific.SpecificRecord;
+import org.apache.avro.util.Utf8;
+import org.apache.flink.types.Row;
+import org.apache.flink.util.Preconditions;
+
+/**
+ * Serialization schema that serializes {@link Row} over {@link SpecificRecord} into a Avro bytes.
+ */
+public class AvroRowSerializationSchema implements SerializationSchema<Row> {
+
+	/**
+	 * Avro serialization schema.
+	 */
+	private final Schema schema;
+
+	/**
+	 * Writer to serialize Avro record into a byte array.
+	 */
+	private final DatumWriter<GenericRecord> datumWriter;
+
+	/**
+	 * Output stream to serialize records into byte array.
+	 */
+	private final ByteArrayOutputStream arrayOutputStream =  new ByteArrayOutputStream();
+
+	/**
+	 * Low-level class for serialization of Avro values.
+	 */
+	private final Encoder encoder = EncoderFactory.get().binaryEncoder(arrayOutputStream, null);
+
+	/**
+	 * Creates a Avro serialization schema for the given schema.
+	 *
+	 * @param recordClazz Avro record class used to deserialize Avro's record to Flink's row
+	 */
+	@SuppressWarnings("unchecked")
+	public AvroRowSerializationSchema(Class<? extends SpecificRecord> recordClazz) {
+		Preconditions.checkNotNull(recordClazz, "Avro record class must not be null.");
+		this.schema = SpecificData.get().getSchema(recordClazz);
+		this.datumWriter = new ReflectDatumWriter<>(schema);
+	}
+
+	@Override
+	@SuppressWarnings("unchecked")
+	public byte[] serialize(Row row) {
+		// convert to record
+		final Object record = convertToRecord(schema, row);
+
+		// write
+		try {
+			arrayOutputStream.reset();
+			datumWriter.write((GenericRecord) record, encoder);
+			encoder.flush();
+			return arrayOutputStream.toByteArray();
+		} catch (IOException e) {
+			throw new RuntimeException("Failed to serialize Row.", e);
+		}
+	}
+
+	/**
+	 * Converts a (nested) Flink Row into Avro's {@link GenericRecord}.
+	 * Strings are converted into Avro's {@link Utf8} fields.
+	 */
+	private static Object convertToRecord(Schema schema, Object rowObj) {
+		if (rowObj instanceof Row) {
+			// records can be wrapped in a union
+			if (schema.getType() == Schema.Type.UNION) {
+				final List<Schema> types = schema.getTypes();
+				if (types.size() == 2 && types.get(0).getType() == Schema.Type.NULL && types.get(1).getType() == Schema.Type.RECORD) {
+					schema = types.get(1);
+				}
+				else {
+					throw new RuntimeException("Currently we only support schemas of the following form: UNION[null, RECORD]. Given: " + schema);
+				}
+			} else if (schema.getType() != Schema.Type.RECORD) {
+				throw new RuntimeException("Record type for row type expected. But is: " + schema);
+			}
+			final List<Schema.Field> fields = schema.getFields();
+			final GenericRecord record = new GenericData.Record(schema);
```

Comment:
Those could be reused as well (and recreated if `null` at some point).
I think it would be possible, but a bit fiddly.
So let's keep it as it is for now.